### PR TITLE
Flexible GPU and CPU Usage for Gravity

### PR DESCRIPTION
**Pull request by spencerw**
Although calculating gravity on the GPU is generally much faster, doing so introduces a significant amount of extra overhead. When using multistepping, it's often the case that using the GPU for gravity on higher rungs actually slows things down.

This PR introduces a new parameter 'nGpuMinParts', which triggers a tree walk on the CPU (instead of the GPU) whenever the number of active particles falls below the threshold.

### CODE REVIEW

**Comment by spencerw** [comment]
Note that there are still a couple of 'todo' items before this can be merged. Most importantly, enabling the periodic force calculation currently causes a hang whenever the gravity tree walk happens on the CPU. @trquinn I was hoping you might have some ideas as to why that could be happening.

**Comment by spencerw** [comment]
The hang during the gravity step was caused by a misplaced #ifdef inside of calculateEwald and is now fixed.

Some of the 'useckloop' code still needs to be incorporated and there are a couple of places where indentation needs to be updated, but a code review would definitely be helpful at this point. In particular, this PR involved creating a considerable amount of duplicated code, to ensure that the CPU tree walk can be done with or without the CUDA compile flag. Moving things into helper functions cuts down on this, but it's definitely made the code messier and harder to read.

**Comment by spencerw** [comment]
@trquinn Are we sure that removing 'bUseCpu' as a member from the TreePiece and ListCompute classes is what we want to do? I see at least 15 function headers that will need to be changed if we instead pass it down as an argument from Main::startGravity to all of the places it's needed.

**Comment by trquinn** [comment]
I think it's find to keep the bUseCpu as Treepiece and ListCompute attributes.  I recommend NOT putting them under an `#ifdef CUDA`

File: Compute.h [hunk 1]
```diff
@@ -179,6 +183,7 @@ class ListCompute : public Compute{
   State *getNewState();
   void freeState(State *state);
   void freeDoubleWalkState(DoubleWalkState *state);
+  void enableCpu() {bUseCpu = 1;}
```
Comment (trquinn) [inline Compute.h:186]: Document

File: TreePiece.cpp [hunk 2]
```diff
@@ -3749,20 +3749,26 @@ void TreePiece::finishBucket(int iBucket) {
       }
 
 #if defined CUDA
```
Comment (trquinn) [inline TreePiece.cpp:3751]: Can't this ifdef now go away?

File: TreePiece.cpp [hunk 3]
```diff
@@ -3861,6 +3867,15 @@ void TreePiece::continueWrapUp(){
   }
 }
 
+// Schedule a walk on the CPU
```
Comment (trquinn) [inline TreePiece.cpp:3870]: Make this DOXYGEN style.

File: TreePiece.cpp [hunk 4]
```diff
@@ -3872,33 +3887,32 @@ void TreePiece::doAllBuckets(){
 #endif
 
 #ifdef GPU_LOCAL_TREE_WALK
-  ListCompute *listcompute = (ListCompute *) sGravity;
-  DoubleWalkState *state = (DoubleWalkState *)sLocalGravityState;
+  if (!bUseCpu) {
+    ListCompute *listcompute = (ListCompute *) sGravity;
+    listcompute->enableCpu();
```
Comment (trquinn) [inline TreePiece.cpp:3892]: This is non-obvious.  why set the Cpu flag here?  Is it for the next iteration?  If so, add a comment.

File: TreePiece.cpp [hunk 5]
```diff
@@ -4144,7 +4152,26 @@ void TreePiece::calculateEwald(EwaldMsg *msg) {
   } else {
     delete msg;
   }
+}
+
+/// @brief Start the ewald calculation on this TreePiece
+/// @param msg Indicates whether this function was called from EwaldInit
+void TreePiece::calculateEwald(EwaldMsg *msg) {
+#ifdef SPCUDA
```
Comment (trquinn) [inline TreePiece.cpp:4160]: Can this ifdef now be deleted?

File: Compute.cpp [hunk 6]
```diff
@@ -1547,354 +1567,360 @@ void ListCompute::stateReady(State *state_, TreePiece *tp, int chunk, int start,
     CkPrintf("memcheck before list iteration\n");
     CmiMemoryCheck();
 #endif
-#ifndef CUDA
-  for(int b = start; b < end; b++){
-    if(tp->bucketList[b]->rungs >= activeRung){
 
-      for(int level = 0; level <= maxlevel; level++){
+#ifdef CUDA
```
Comment (trquinn) [inline Compute.cpp:1571]: Can we get rid of this ifdef?

**Review by trquinn — CHANGES_REQUESTED** [review]
Some comment cleanup and possible removal of some ifdefs.

File: Compute.cpp [hunk 7]
```diff
@@ -1548,352 +1604,324 @@ void ListCompute::stateReady(State *state_, TreePiece *tp, int chunk, int start,
...
-        if(hasLocalLists){
-          CkVec<LocalPartInfo> &lpilist = state->lplists[level];
-          computed = calcParticleForces(tp, b, activeRung, lpilist);
-          tp->addToParticleInterLocal(computed);
-        }
-      }// level
-
-    }// active
-  }// bucket
+  helper(state, tp, chunk, start, end, maxlevel, hasRemoteLists, hasLocalLists);
```
[the remark this answers is not in the document]
Comment (spencerw) [reply Compute.cpp]: This was changed in [5f7275e](https://github.com/N-BodyShop/changa/pull/170/commits/5f7275eab9dfb205769120ed4389ef42250d2809), along with any other '#ifdef CUDA, #else' blocks that contained repeated code.

File: ParallelGravity.h [hunk 8]
```diff
@@ -908,6 +908,7 @@ class TreePiece : public CBase_TreePiece {
         size_t sCompactParts;
         size_t sVarParts;
 	cudaStream_t stream;
+        int bUseGpu;
```
[the remark this answers is not in the document]
Comment (spencerw) [reply ParallelGravity.h]: Done. Note that this variable has been changed to 'bUseCpu' and was moved to ParallelGravity.h:1204 (see [5f7275e](https://github.com/N-BodyShop/changa/pull/170/commits/5f7275eab9dfb205769120ed4389ef42250d2809))

File: ParallelGravity.h [hunk 9]
```diff
@@ -1412,6 +1413,7 @@ class TreePiece : public CBase_TreePiece {
 	/** @brief Start a full step of bucket computation, it sends a message
 	 * to trigger nextBucket() which will loop over all the buckets.
 	 */
+       void schedCpuWalk();
```
[the remark this answers is not in the document]
Comment (spencerw) [reply ParallelGravity.h:1418]: Using the #ifdef trick, there is no longer any repeated code in TreePiece::doAllBuckets, so this function has been removed.

File: TreePiece.cpp (hunk 2 above)
Comment (spencerw) [reply TreePiece.cpp:3751]: Doesn't this need to stay as is? If building with CUDA and bUseCpu is false, we want to make sure continueWrapUp() doesn't trigger.

File: TreePiece.cpp (hunk 4 above)
Comment (spencerw) [reply TreePiece.cpp:3892]: Extraneous call from when I was debugging. It has been removed.

**Comment by spencerw** [comment]
All of the requested changes have been made. Note that the review was started before my previous commit ([5f7275e](https://github.com/N-BodyShop/changa/pull/170/commits/5f7275eab9dfb205769120ed4389ef42250d2809)), so some of the change requests are marked as 'outdated'.

It sounds like you noticed the new commit (implementing the ifdef trick to get rid of duplicated code) partway through, however.

**Review by trquinn — CHANGES_REQUESTED** [review]
One change is correcting a comment, the other is a bug, although on a code path that is rarely used.

Also: I'm getting memory errors on "teststep" when running on the A100s with 1 node, 2 processes (i.e. 2 GPUs), 2 PEs/process.  In short
```
ibrun -n 2 ../ChaNGa.smp.cuda ++ppn 2 +setcpuaffinity +commap 0,8 +pemap 1-2,9-10 ./test_pg.param
```
The error is
```
Calculating gravity (tree bucket, theta = 0.700000) ... Fatal CUDA Error an illegal memory access was encountered at HostCUDA.cu:609.
Return value 700 from 'cudaFree(ptr->d_list)'.------------- Processor 3 Exiting: Caught Signal ------------
Reason: Aborted
Fatal CUDA Error an illegal memory access was encountered at HostCUDA.cu:588.
Return value 700 from 'cudaMalloc(&ptr->d_bucketMarkers, markerSize)'.[v330-020.ls6.tacc.utexas.edu:mpispawn_0][readline] Unexpected En
d-Of-File on file descriptor 5. MPI process died?
```

**Review by trquinn — APPROVED** [review]

**Pull request merged**